### DISCUSSION TITLE
-monochrome command no longer dithers grayscale images

### DISCUSSION THREAD

**Discussion by nikitos1979** [ImageMagick/ImageMagick#5156] [Help]
We use IM to convert grayscale images to B/W.  Up to version 7.0.11-12, -monochrome would automatically dither the image, however since then the dithering no longer occurs.  We have tried various -ordered-dither options but nothing comes close to the dithering we saw with the older -monochrome option.

[images.zip](https://github.com/ImageMagick/ImageMagick/files/8702528/images.zip)

**Comment by fmw42** [comment]
I can confirm the issue with IM 6.9.12-49 Q16 and 7.1.0-34 Q16-HDRI. The result seems to be a simple binary threshold

```
convert original.tif -monochrome x.tif

magick original.tif -monochrome x.tif

```


[x.tif.zip](https://github.com/ImageMagick/ImageMagick/files/8702792/x.tif.zip)

**Comment by urban-warrior** [comment]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/download/beta/ by sometime tomorrow.

**Reply by nikitos1979** [reply]
Thank you!

**Comment by nikitos1979** [comment]
How long does the Beta last? I tried version 7.1.0.37 and the issue still persists there.

**Comment by urban-warrior** [comment]
It appears that modern ImageMagick is doing such a good job reducing the colors, there is little error to distribute to produced the speckles.  @fmw42, can you reproduce the previous results (see images.zip above)?

**Comment by fmw42** [comment]
These two command on the current IM 6.9.12-52 and IM 7.1.0-37 

```
convert original.tif -monochrome x6.tif

magick original.tif -monochrome x7.tif

```

produce bilevel images, but they are not dithered per https://imagemagick.org/Usage/quantize/#monochrome.

Perhaps one might make -monochrome sensitive to +dither and -dither.

**Reply by fmw42** [reply]
I just noticed that original.tif (x.tif in the zip file) was already bi-level. So my tests are not very relevant until the OP posts some grayscale input that is not bi-level already.

**Reply by fmw42** [reply]
Sorry, I got the wrong zip file from the post. I will retest with the correct original.tif that is grayscale and report back.

**Comment by fmw42** [comment]
OK. Using the correct grayscale original.tif, I get for these two command on the current IM 6.9.12-52 and IM 7.1.0-37

```
convert original.tif -monochrome x6.tif

magick original.tif -monochrome x7.tif
```
reasonable looking bilevel images, but they are not dithered per https://imagemagick.org/Usage/quantize/#monochrome. So they do not look like the old version of -monochrome which keeps the gray-looking result similar to the input, but which is actually dithered bilevel. 

Perhaps one might make -monochrome sensitive to +dither and -dither. If -dither, then dither as we used to do. If +dither, then keep it the way you have without dithering.

**Comment by carrchr1974** [comment]
Has the fix been proven successful in beta and if so when will it moved to production?

**Reply by nikitos1979** [reply]
No, and I have not found a good workaround, so we have not been able to upgrade past version 7.0.11-12

**Comment by rodrigo-werneck** [comment]
No solution yet? Isn´t there anyone addressing that issue anymore?

**Comment by fmw42** [comment]
The options page for -dither still says:

> The color reduction operators [-colors](https://imagemagick.org/script/command-line-options.php#colors), [-monochrome](https://imagemagick.org/script/command-line-options.php#monochrome), [-remap](https://imagemagick.org/script/command-line-options.php#remap), and [-posterize](https://imagemagick.org/script/command-line-options.php#posterize), apply dithering to images using the reduced color set they created.

which includes -monochrome.   

I tried adding -dither Floyd-Steinberg to the command before -monochrome, but that did not help.

**Reply by rodrigo-werneck** [reply]
So did I.
Shall we give up ImageMagick and assess its competitors?

**Reply by urban-warrior** [reply]
If ImageMagick does not fit your requirements, we recommend you seek out an alternative solution.

**Reply by rodrigo-werneck** [reply]
It´s really a pity as it fitted during a long time in the past. Perhaps this loss is one of the reasons why repositories like Solaris OpenCSW kept an old version (7.0.8).

**Comment by fmw42** [comment]
@urban-warrior, perhaps you might explain the reason that -monochrome no longer supports the option for dithered graphics.  Would it not be easy to add a define or make it sensitive to -dither.  Dithered results seemed to be popular for this function.

**Reply by urban-warrior** [reply]
Its an issue that we will visit some day and perhaps fix.  We just don't have the time to do so right now.

**Reply by fmw42** [reply]
Thanks for the clarification.

**Comment by rodrigo-werneck** [comment]
> We use IM to convert grayscale images to B/W. Up to version 7.0.11-12, -monochrome would automatically dither the image, however since then the dithering no longer occurs. We have tried various -ordered-dither options but nothing comes close to the dithering we saw with the older -monochrome option.
> 
> [images.zip](https://github.com/ImageMagick/ImageMagick/files/8702528/images.zip)

Doing some tests with 7.0.11-12, it seems to be doing some dithering, but not as good as 7.0.8-29.
At which versions between those two has changes been made in this topic?

**Reply by fmw42** [reply]
As stated above -monochrome does not dither in IM 7

**Reply by rodrigo-werneck** [reply]
Maybe. But 7.01.11-12 delivers smoothier output than 7.1.1-12.
And if 7.0.8-29 doesn´t do dithering in -monochrome mode (which is the case in all my tests), I really don´t know what dithering is at all.

**Reply by fmw42** [reply]
The code for -monochrome was modified several times. For dithering of monochrome in IM 6 see https://imagemagick.org/Usage/quantize/#monochrome

**Comment by malosupremo** [comment]
I can confirm that even with the latest version 7.1.1-30 the -monocrhome does not dither a grayscale image, but applies a threshold, and that's not what the documentation says

**Reply by nikitos1979** [reply]
I'm not holding my breath for this to be fixed.  I ended up using a custom dithering algorithm (Stucky).  You can look for open source code for this on the web.

**Reply by malosupremo** [reply]
for sure; anyway I partially solved using "-dither FloydSteinberg -remap pattern:gray50 -colors 2 -normalize" ... but it's far from an optimal solution. Thanks for the suggestion, I'll look it up

**Reply by nikitos1979** [reply]
FloydSteinberg is an old algorithm, they developed better ones.  Check out this article:  https://tannerhelland.com/2012/12/28/dithering-eleven-algorithms-source-code.html

**Reply by fmw42** [reply]
You can create your own dither pattern in Imagemagick. See https://imagemagick.org/Usage/quantize/#diy_dither

**Comment by snibgo** [comment]
The link from fmw42 doesn't show how to incorporate error diffusion into our own dithering. I think that would need low-level coding, to read the patterns shown in the link from nikitos1979. Then we could experiment with Sierra dithering, and diffusing error over more than just a couple of pixels, and so on.

I'm currently to busy to do this.

**Comment by fmw42** [comment]
What version of Imagemagick 6 was used to create the previous_monochrome.TIF provided by the OP (@nikitos1979)?

**Reply by nikitos1979** [reply]
Any version 7 build prior to 7.0.11-12

**Reply by fmw42** [reply]
No sorry, I tried convert on IM 6.9.10.72 and 6.9.12.13 and neither was not dithered.  So what was the exact version of IM 6 that you used to get the dither you show?

**Reply by nikitos1979** [reply]
I don't know which version of IM 6.x dithered or not.  I do know that all versions of IM 7.x dithered prior to 7.0.11-12.

**Comment by fmw42** [comment]
Here is an attempt to reproduce -monochrome. It is not perfect but may suffice.

```
magick \
\( xc:black xc:white +append -write mpr:map +delete \) \
original.tif -remap mpr:map -contrast-stretch 1x10% \
original_remap_dither.gif

```

![original_remap_dither](https://github.com/ImageMagick/ImageMagick/assets/13458218/ccfaad9b-5c1e-4e54-b308-45493fd0dcc7)

**Comment by urban-warrior** [comment]
Any dithering you've experienced in the past resulted from a flawed algorithm. Monochrome employs a color reduction algorithm to reduce N colors to 2, while dithering disperses any error between the original and quantized colors among their neighbors. However, since the original image is already monochrome, color reduction becomes effectively redundant. There is no error to distribute to neighboring pixels.

**Comment by karme** [comment]
hmm really unfortunate incompatibility between imagemagick 6 and 7 :-(

**Reply by karme** [reply]
i now use something like:

convert ... pgm:-|pamditherbw|convert pam:- ...
